deepcell/applications/segmentation_application.py

```diff
@@ -0,0 +1,140 @@
+            model_image_shape (tuple, optional): Shape of input expected by model.
+                Defaults to (128, 128, 1).
+            dataset_metadata (optional): Any input, e.g. str or dict. Defaults to None.
+            model_metadata (optional): Any input, e.g. str or dict. Defaults to None.
+            model_mpp (float, optional): Microns per pixel resolution of training data.
+                Defaults to 0.65.
+            preprocessing_fn (function, optional): Preprocessing function to apply to data
+                prior to prediction. Defaults to None.
+            postprocessing_fn (function, optional): Postprocessing function to apply
+                to data after prediction. Defaults to None.
+                Must accept an input of a list of arrays and then return a single array.
+        """
+
+        self.model = model
+
+        self.model_image_shape = kwargs.get('model_image_shape', (128, 128, 1))
+        self.model_mpp = kwargs.get('model_mpp', 0.65)
+        self.preprocessing_fn = kwargs.get('preprocessing_fn', None)
+        self.postprocessing_fn = kwargs.get('postprocessing_fn', None)
+        self.dataset_metadata = kwargs.get('dataset_metadata', None)
+        self.model_metadata = kwargs.get('model_metadata', None)
+
+    def predict(self, image,
+                batch_size=4,
+                image_mpp=None,
+                preprocess_kwargs={},
+                postprocess_kwargs={},
+                debug=False):
+        """Generates a labeled image of the input running prediction with
+        appropriate pre and post processing functions.
+
+        Input images are required to have 4 dimensions [batch, x, y, channel]. Additional
+        empty dimensions can be added using `np.expand_dims`
+
+        Args:
+            image (np.array): Input image with shape [batch, x, y, channel]
+            batch_size (int, optional): Number of images to predict on per batch. Defaults to 4.
+            image_mpp (float, optional): Microns per pixel for the input image. Defaults to None.
+            preprocess_kwargs (dict, optional): Kwargs to pass to preprocessing function.
+                Defaults to {}.
+            postprocess_kwargs (dict, optional): Kwargs to pass to postprocessing function.
+                Defaults to {}.
+            debug (bool, optional): If True, returns intermediate outputs of data processing.
+                Defaults to False.
+
+        Raises:
+            ValueError: Input data must have 4 dimensions, [batch, x, y, channel
+
+        Returns:
+            np.array: Labeled image, if debug is False.
+                If debug is True, returns (image, tiles, output_tiles, output_images, label_image)
+        """
+
+        # Check input size of image
+        if len(image.shape) != 4:
+            raise ValueError('Input data must have 4 dimensions, [batch, x, y, channel'
+                             'Input data only has {} dimensions'.format(str(len(image.shape))))
```

Comment:
Instead of `4`, maybe we could use a more general solution, like `required_rank`.

Or get it out of the `model.input_shape`?

Reply:
Changed to compare to `self.required_rank` defined as `len(self.model_input_shape) + 1`